datumaro/components/dataset.py

```diff
@@ -383,6 +381,11 @@ def _update_status(item_id, new_status: ItemStatus):
 
         i = -1
         for i, item in enumerate(source):
+            if self.media_type():
+                if item.media and not isinstance(item.media, self.media_type()):
+                    raise MediaTypeError("Dataset elements must have a '%s' " \
+                        "media type" % self.media_type())
```

Comment:
Probably, it is reasonable to deduce and cache this info during processing, if it was not specified on creation.